include/circt/Dialect/HW/HWSymCache.h

```diff
@@ -59,9 +62,22 @@ class SymbolCache {
     symbolCache.try_emplace(name, op, port);
   }
 
+  /// Populate the symbol cache with all symbol-defining operations within the
+  /// 'top' module op.
+  void addDefinitions(mlir::ModuleOp top) {
+    for (mlir::Operation &op : top.getOps())
+      if (auto symOp = llvm::dyn_cast<mlir::SymbolOpInterface>(op))
+        if (auto name = symOp.getNameAttr())
+          addDefinition(name, symOp);
+  }
+
   /// Mark the cache as frozen, which allows it to be shared across threads.
   void freeze() { isFrozen = true; }
 
+  /// Mark the cache as unfrozen, allowing for mutation. Caller should ensure
+  /// that the cache is no longer accessed across threads.
+  void unfreeze() { isFrozen = false; }
```

Comment:
If you need to make changes to HWSymCache, please break them out for separate review. Notably, I think there will be strong opinions about `unfreeze`, as it effectively removes the two-phase guarantee about this utility (replacing it with n-phases where n is even).

Or, another way of putting it: can you make it work without `unfreeze`? The `addDefinitions` method seems "obvious" and something you could just commit right now, but if you really need to add `unfreeze` I think it should be discussed separately from parameterization.